### PR TITLE
[FLINK-17148][python] Support converting pandas DataFrame to Flink Table

### PR DESCRIPTION
**Pull request by dianfu**
## What is the purpose of the change

*This pull request add the support for converting pandas dataframe to flink table.*

## Brief change log

  - *Introduce ArrowSourceFunction and ArrowTableSource which takes the serialized byte array of arrow record batch as the source data*
  - *Add TableEnvironment.from_pandas which could be used to convert pandas dataframe to flink table*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added Java tests ArrowSourceFunctionTest and RowArrowSourceFunctionTest*
  - *Added Python tests test_pandas_conversion.py*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-python/src/test/java/org/apache/flink/table/runtime/arrow/sources/ArrowSourceFunctionTestBase.java [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+		});
+		runner2.start();
+
+		if (!latch.isTriggered()) {
+			latch.await();
+		}
+		runner2.join();
+
+		Assert.assertNull(error[0]);
+		Assert.assertEquals(testData.f0.size(), numOfEmittedElements.get());
```
Comment (hequn8128) [inline flink-python/src/test/java/org/apache/flink/table/runtime/arrow/sources/ArrowSourceFunctionTestBase.java:147]: Also verify the content of the data?

File: flink-python/src/test/java/org/apache/flink/table/runtime/arrow/sources/ArrowSourceFunctionTestBase.java [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+		}
+
+		runner.join();
+		runner2.join();
+		testHarness.close();
+		testHarness2.close();
+
+		Assert.assertNull(error[0]);
+		Assert.assertNull(error[1]);
+		Assert.assertEquals(testData.f0.size(), numOfEmittedElements.get());
```
Comment (hequn8128) [inline flink-python/src/test/java/org/apache/flink/table/runtime/arrow/sources/ArrowSourceFunctionTestBase.java:215]: Also verify the content of the data?

File: flink-python/src/test/java/org/apache/flink/table/runtime/arrow/sources/ArrowSourceFunctionTestBase.java [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+			}
+		};
+
+		// run the source asynchronously
+		Thread runner = new Thread(() -> {
+			try {
+				arrowSourceFunction.run(sourceContext);
+			} catch (Throwable t) {
+				if (!t.getMessage().equals("Fail the arrow source")) {
+					error[0] = t;
```
Comment (hequn8128) [inline flink-python/src/test/java/org/apache/flink/table/runtime/arrow/sources/ArrowSourceFunctionTestBase.java:99]: Add the corresponding assert to verify that error[0] is not null?

**Comment by hequn8128** [comment]
BTW. Would be great if you can rebase to the master. The interface in BaseRow has been changed, i.e., `getHeader()` has been replaced with `getRowKind()`.

File: flink-python/pyflink/table/tests/test_pandas_conversion.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+from pyflink.table.types import DataTypes, Row
+from pyflink.testing import source_sink_utils
+from pyflink.testing.test_case_utils import PyFlinkBlinkBatchTableTestCase, \
+    PyFlinkBlinkStreamTableTestCase, PyFlinkStreamTableTestCase
+
+
+class PandasConversionTestBase(object):
+
+    @classmethod
+    def setUpClass(cls):
```
[the remark this answers is not in the document]
Comment (dianfu) [reply flink-python/pyflink/table/tests/test_pandas_conversion.py:30]: The name `setUpClass` is from unittest.TestCase and I guess we can not change it.

File: flink-python/pyflink/table/tests/test_pandas_conversion.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                  DataTypes.FIELD("c", DataTypes.TIMESTAMP(3)),
+                  DataTypes.FIELD("d", DataTypes.ARRAY(DataTypes.INT()))]))])
+        cls.pdf = cls.create_pandas_data_frame()
+
+    @classmethod
+    def create_pandas_data_frame(cls):
+        data_dict = {}
+        for j, name in enumerate(cls.data_type.names):
+            data_dict[name] = [cls.data[i][j] for i in range(len(cls.data))]
+        # need convert to numpy types
```
[the remark this answers is not in the document]
Comment (dianfu) [reply flink-python/pyflink/table/tests/test_pandas_conversion.py:71]: The integer types will be parsed as `int64` by default and so we need to specify it explicitly.

File: flink-python/src/test/java/org/apache/flink/table/runtime/arrow/sources/ArrowSourceFunctionTestBase.java (hunk 3 above)
Comment (dianfu) [reply flink-python/src/test/java/org/apache/flink/table/runtime/arrow/sources/ArrowSourceFunctionTestBase.java:99]: error[0] should always be null and it has been asserted at the end of the test. I'm not sure what do you mean about `verify that error[0] is not null`?

Comment (hequn8128) [reply flink-python/src/test/java/org/apache/flink/table/runtime/arrow/sources/ArrowSourceFunctionTestBase.java:99]: Ignore my comment here. You are right.

File: flink-python/pyflink/table/tests/test_pandas_conversion.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        self.t_env.execute("test")
+        actual = source_sink_utils.results()
+        self.assert_equals(actual,
+                           ["1,1,1,1,true,1.1,1.2,hello,[97, 97, 97],"
+                            "1000000000000000000.010000000000000000,2014-09-13,01:00:01,"
+                            "1970-01-01 00:00:00.123,[hello, 中文],1,hello,"
+                            "1970-01-01 00:00:00.123,[1, 2]"])
+
+
+class StreamPandasConversionTests(PandasConversionITTests,
```
[the remark this answers is not in the document]
Comment (dianfu) [reply flink-python/pyflink/table/tests/test_pandas_conversion.py:134]: Most code could not be reusable for the batch mode of the old planner. So I'd like to handle it in a separate JIRA if needed.

**Comment by dianfu** [comment]
@hequn8128 Thanks a lot for your valuable feedback. I have updated the PR accordingly.

File: docs/dev/table/python/conversion_of_pandas.zh.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+{% highlight python %}
+import pandas as pd
+import numpy as np
+
+# Create a Pandas DataFrame
+pdf = pd.DataFrame(np.random.rand(1000, 2))
+
+# Create a PyFlink Table from a Pandas DataFrame
+table = t_env.from_pandas(pdf)
```
Comment (hequn8128) [inline docs/dev/table/python/conversion_of_pandas.zh.md:47]: Maybe add more examples here. For example, how to specify table names, which is commonly required.

**Comment by dianfu** [comment]
@hequn8128 Thanks for the review. Updated.

**Comment by dianfu** [comment]
Not sure why the CI of Azure wasn't triggered. It has succeed in my private azure pipeline: https://dev.azure.com/dianfu/Flink/_build/results?buildId=50&view=results

**Comment by hequn8128** [comment]
@dianfu Thanks. Merging...

**Pull request merged**